### ISSUE TITLE
Msf::Post::File.read_file will fail when session.type is not meterpreter/shell

### ISSUE THREAD

**Issue by bcoles** [rapid7/metasploit-framework#12584]
Once upon a time, session types were only `meterpreter` or `shell`. Now, `powershell` is a supported `session.type` (and potentially `ssh` in the future).

`Msf::Post::File.read_file` will immediately return `nil` if the `session.type` is not `meterpreter` or `shell`.

```ruby
  #
  # Platform-agnostic file read.  Returns contents of remote file +file_name+
  # as a String.
  #
  # @param file_name [String] Remote file name to read
  # @return [String] Contents of the file
  #
  # @return [Array] of strings(lines)
  #
  def read_file(file_name)
    if session.type == 'meterpreter'
      return _read_file_meterpreter(file_name)
    end

    return nil unless session.type == 'shell'

    if session.platform == 'windows'
      return session.shell_command_token("type \"#{file_name}\"")
    end

    return nil unless readable?(file_name)

    if command_exists?('cat')
      return session.shell_command_token("cat \"#{file_name}\"")
    end

    # Result on systems without cat command
    session.shell_command_token("while read line; do echo $line; done <#{file_name}")
  end
```

**Comment by bcoles** [comment]
Adding `bug` label. Although this hasn't been confirmed, it is most likely a bug.

**Comment by bcoles** [comment]
Using the same approach as `write_file` to check `session.respond_to? :shell_command_token` may or may not be a viable alternative.

```ruby
  # Platform-agnostic file write. Writes given object content to a remote file.
  #
  # NOTE: *This is not binary-safe on Windows shell sessions!*
  #
  # @param file_name [String] Remote file name to write
  # @param data [String] Contents to put in the file
  # @return [void]
  def write_file(file_name, data)
    if session.type == "meterpreter"
      fd = session.fs.file.new(file_name, "wb")
      fd.write(data)
      fd.close
    elsif session.respond_to? :shell_command_token
      if session.platform == 'windows'
        session.shell_command_token("echo #{data} > \"#{file_name}\"")
      else
        _write_file_unix_shell(file_name, data)
      end
    end
    true
  end
```

As an aside, the `write_file` method appears to also contain a bug. If the `session.type` is not `meterpreter` and the `session` does not `respondto? :shell_command_token`, then the method will do nothing and return `true`. That does not appear to be intended, and likely unexpected behaviour.

**Comment by bcoles** [comment]
Fixed in #15320.
